M2 blog (#139)
eigerco/move-spec-testing: m2.md

```bash
move-mutation-test run --output report.json --operators delete_statement,binary_operator_replacement,if_else_replacement
```

Available operators: `unary_operator_replacement`, `delete_statement`, `break_continue_replacement`, `binary_operator_replacement`,
`if_else_replacement`, `literal_replacement`, `binary_operator_swap`.

**Note:** The `--mode` and `--operators` options are mutually exclusive.

### Performance

To speed up mutation testing, we added a `--fail-fast` option to `move-mutation-test`. This stops test execution as soon as a test fails
(indicating that the mutant was caught). The default value is `true`, but it can be turned off if desired.

## Case Study

We ran the tool on a large Move project, splitting up the test into 3 runs, light, medium-only and heavy-only.

| Mode        | Mutants Alive | Mutants Killed | Percentage Alive |
| ----------- | ------------- | -------------- | ---------------- |
| light       | 88            | 353            | 20 %             |
| medium-only | 1155          | 1458           | 44 %             |
| heavy-only  | 1088          | 3952           | 22 %             |

### Examples per operator

> As there are in total over 2000 surviving mutants, we can not cover them all. A few examples are given per operator.

Reviewer: ```suggestion
The following chapter presents a selection of real, surviving mutants from actual projects, with a few examples provided per operator, as it is not feasible to cover all 2000+ mutants.
```

```

Another example, here the tests should assert that the remaining size is correctly calculated:

```move
async_matching_engine::trigger_pending_twap: Alive mutant
--- original
+++ modified
@@ -840,7 +840,7 @@
                 account,
                 order_id,
                 orig_size,
-                remaining_size: twap_size - total_fill_size,
+                remaining_size: total_fill_size - twap_size,
                 is_buy,
                 is_reduce_only,
                 twap_start_time_s,
```

#### delete_statement

This operator deletes entire statements.

In this example, tests should assert that this move to global storage has happened.

```move
position_tp_sl_tracker::register_market: Alive mutant
--- original
+++ modified
@@ -56,13 +56,7 @@
     }

     public(package) fun register_market(market_signer: &signer) {
-        move_to(
-            market_signer,
-            PendingOrderTracker {
-                price_move_up_index: new_default_big_ordered_map(),
-                price_move_down_index: new_default_big_ordered_map()
-            }
-        )
+        {}
     }

     public(package) fun get_account_from_pending_request(

> 

Reviewer: I guess this is part of the original diff snippet, but we could remove it for better clarity in the blog post.
```suggestion
```

--- original
+++ modified
@@ -91,7 +91,7 @@
         let key = BuilderAndAccount { account: user_address, builder };
         let approved_max_fees = registry.approved_max_fees.get(&key);
         if (approved_max_fees.is_none()) {
-            abort error::invalid_argument(EINSUFFICIENT_BALANCE);
+            {};
         };
         registry.approved_max_fees.remove(&key);
     }
```

#### break_continue_replacement

This mutation operator replaces or deletes break/continue statements with other break/continue statements.

If this break statement were lost, the semantics would probably no longer be valid.

```move
pending_order_tracker::remove_reduce_only_order: Alive mutant
--- original
+++ modified
@@ -321,7 +321,7 @@
             if (&reduce_only_orders.orders[i].order_id == &order_id) {
                 reduce_only_orders.total_size -= reduce_only_orders.orders[i].size;
                 reduce_only_orders.orders.remove(i);
-                break;
+                {};
             };
             i += 1
         };
```

#### literal_replacement

This mutation operator replaces literals with other literals. For example, the 0 literal can be replaced

> with the 1 literal or other random literal, true to false, etc. This mutation operator can also replace addresses with other addresses.

Reviewer: ```suggestion
This mutation operator replaces literals with other literals. For example, the `0` literal can be replaced
with the `1` literal or other random literal, `true` to `false`, etc. This mutation operator can also replace addresses with other addresses.
```

> with the 1 literal or other random literal, true to false, etc. This mutation operator can replace also addresses with other addresses.

Reviewer 2: ```suggestion
with the 1 literal or other random literal, true to false, etc. This mutation operator can also replace addresses with other addresses.
```
